File: implementations/typescript/provekit-realize-typescript-better-sqlite3/src/platform_semantics.js
// SPDX-License-Identifier: Apache-2.0
//
// Binding-kit platform semantics declaration for the TypeScript better-sqlite3 library.
//
// Implements the provekit.plugin.platform_semantics RPC method (PEP 1.7.0).
// Returns the JSON payload for the "typescript-better-sqlite3" binding target.
//
// RowIdMechanism = LastInsertRowid: the library exposes the inserted row id
// as `stmt.run(...).lastInsertRowid`, which reads connection-global mutable
// state maintained by SQLite and set by the most recent INSERT on that
// connection. This is structurally different from the PostgreSQL mechanism
// (RETURNING clause) and requires a distinct migration pattern.
//
// CID computation follows the substrate spec:
//   DimensionValueMemento CID = blake3-512(JCS(memento WITHOUT cid + kit_cid))
//   PlatformSemanticTag CID   = blake3-512(JCS(tag WITHOUT cid + kit_cid))

"use strict";

const { blake3 } = require("@noble/hashes/blake3.js");

const KIT_ID = "provekit-binding-better-sqlite3@0.1.0";

// CID for concept:insert-and-get-id, minted from its AlgorithmMemento via JCS+blake3-512.
const CONCEPT_INSERT_AND_GET_ID_CID =
  "blake3-512:0a4f0a8d36d8dee96b8d5b32a18bb390f35877ecef611771048c6e10cfc3d25ad8f59de89b00c7794f62cabaf91dbd779244338393a8bb6ef5e8309b0929b3ca";

// kit_cid is provenance metadata only (elided from CID computation per substrate spec).
const KIT_CID = "blake3-512:" + _hexOf(blake3(
  new TextEncoder().encode(KIT_ID),
  { dkLen: 64 }
));

// Cached result to avoid repeated computation.
let _cached = null;

/** Returns the platform_semantics declaration for the better-sqlite3 binding kit. */
function declaration() {
  if (_cached !== null) return _cached;

  const dimValues = dimensionValues();
  const rowIdCid = dimValues.find((d) => d.dimension_name === "RowIdMechanism").cid;

  const tags = [
    _tag(CONCEPT_INSERT_AND_GET_ID_CID, { RowIdMechanism: rowIdCid }),
  ];

  _cached = { tags, dimension_values: dimValues };
  return _cached;
}

/** Returns the dimension values for the better-sqlite3 binding kit. */
function dimensionValues() {
  return [
    // LastInsertRowid: row id is sourced from connection-global mutable state
    // (SQLite last_insert_rowid) set by the most recent INSERT.
    _dimValue("RowIdMechanism", "LastInsertRowid", {
      kind: "atomic",
      name: "row_id_source",
      args: [
        {
          kind: "ctor",
          name: "last_insert_rowid",
          args: [
            { kind: "ctor", name: "connection_state_at_call_return", args: [] },
          ],
        },
      ],
    }),
  ];
}

// --- Helpers ---

function _dimValue(dimensionName, valueName, compareTo) {
  const forCid = {
    compare_to: compareTo,
    dimension_name: dimensionName,
    kind: "platform-dimension-value",
    schemaVersion: "1.0.0",
    value_name: valueName,
  };
  return {
    compare_to: compareTo,
    dimension_name: dimensionName,
    kind: "platform-dimension-value",
    kit_cid: KIT_CID,
    schemaVersion: "1.0.0",
    value_name: valueName,
    cid: _cidOf(forCid),
  };
}

function _tag(opCid, dimensions) {
  const forCid = {
    dimensions,
    kind: "platform-semantic-tag",
    op_cid: opCid,
    schemaVersion: "1.0.0",
  };
  return {
    dimensions,
    kind: "platform-semantic-tag",
    kit_cid: KIT_CID,
    op_cid: opCid,
    schemaVersion: "1.0.0",
    cid: _cidOf(forCid),
  };
}

function _cidOf(obj) {
  const canonical = _jcs(obj);
  const bytes = new TextEncoder().encode(canonical);
  const hash = blake3(bytes, { dkLen: 64 });
  return "blake3-512:" + _hexOf(hash);
}

function _jcs(obj) {
  return JSON.stringify(_sortKeys(obj));
}

function _sortKeys(val) {
  if (val === null || typeof val !== "object") return val;
  if (Array.isArray(val)) return val.map(_sortKeys);
  return Object.fromEntries(
    Object.keys(val).sort().map((k) => [k, _sortKeys(val[k])])
  );
}

function _hexOf(bytes) {
  return Array.from(bytes).map((b) => b.toString(16).padStart(2, "0")).join("");
}

module.exports = { declaration, dimensionValues, CONCEPT_INSERT_AND_GET_ID_CID, KIT_CID };

File: implementations/python/provekit-realize-python-sqlite3/src/provekit_realize_python_sqlite3/platform_semantics.py
from __future__ import annotations

import json
from typing import Any

import blake3 as _blake3

Json = dict[str, Any]

# CID for concept:insert-and-get-id, minted from its AlgorithmMemento via JCS+blake3-512.
CONCEPT_INSERT_AND_GET_ID_CID = (
    "blake3-512:"
    "0a4f0a8d36d8dee96b8d5b32a18bb390f35877ecef611771048c6e10cfc3d25a"
    "d8f59de89b00c7794f62cabaf91dbd779244338393a8bb6ef5e8309b0929b3ca"
)

KIT_ID = "provekit-binding-python-sqlite3@0.1.0"

# kit_cid is provenance metadata only (elided from CID computation per substrate spec).
SQLITE3_KIT_CID = (
    "blake3-512:"
    + _blake3.blake3(KIT_ID.encode("utf-8")).digest(length=64).hex()
)


def dimension_values() -> list[Json]:
    return [
        _with_cid(
            {
                # CursorLastRowid: row id is sourced from the cursor state after
                # executing the INSERT. Accessed as cursor.lastrowid.
                "compare_to": {
                    "kind": "atomic",
                    "name": "row_id_source",
                    "args": [
                        {
                            "kind": "ctor",
                            "name": "cursor_lastrowid",
                            "args": [
                                {
                                    "kind": "ctor",
                                    "name": "cursor_state_after_execute",
                                    "args": [],
                                }
                            ],
                        }
                    ],
                },
                "dimension_name": "RowIdMechanism",
                "kind": "platform-dimension-value",
                "kit_cid": SQLITE3_KIT_CID,
                "schemaVersion": "1.0.0",
                "value_name": "CursorLastRowid",
            }
        )
    ]


def declaration() -> Json:
    dim_vals = dimension_values()
    row_id_cid = next(v["cid"] for v in dim_vals if v["dimension_name"] == "RowIdMechanism")
    return {
        "tags": [
            _with_cid(
                {
                    "dimensions": {"RowIdMechanism": row_id_cid},
                    "kind": "platform-semantic-tag",
                    "kit_cid": SQLITE3_KIT_CID,
                    "op_cid": CONCEPT_INSERT_AND_GET_ID_CID,
                    "schemaVersion": "1.0.0",
                }
            )
        ],
        "dimension_values": dim_vals,
    }


def _with_cid(memento: Json) -> Json:
    # CID is computed over JCS(memento) with "cid" and "kit_cid" elided,
    # per substrate spec (DimensionValueMemento::recompute_cid +
    # PlatformSemanticTag::recompute_cid in provekit-ir-types).
    for_cid = {k: v for k, v in memento.items() if k not in ("cid", "kit_cid")}
    result = dict(memento)
    result["cid"] = _cid_of_json(for_cid)
    return result


def _cid_of_json(value: Any) -> str:
    return (
        "blake3-512:"
        + _blake3.blake3(_canonical_json_bytes(value)).digest(length=64).hex()
    )


def _canonical_json_bytes(value: Any) -> bytes:
    return json.dumps(
        value,
        ensure_ascii=False,
        separators=(",", ":"),
        sort_keys=True,
    ).encode("utf-8")
